File: 2021-09/2021-9-19 650. 只有两个键的键盘/MinSteps.cs
public class Solution {
    public int MinSteps(int n) {
        // dp[1] = 0
        // dp[2] = 2
        // dp[3] = 3
        // dp[4] = 4
        var dp = new int[n+1];
        dp[0] = 0;
        dp[1] = 0;
        for(int i = 2; i <= n; i++)
        {
            dp[i] = int.MaxValue;
            for (int j = 1; j * j <= i; j++)
            {
                if(i % j == 0)
                {
                    dp[i] = Math.Min(dp[i], dp[j] + i / j);
                    dp[i] = Math.Min(dp[i], dp[i / j] + j);
                }
            }
        }

        return dp[n];
    }
}

File: 2021/2021-09/2021-9-19 650. 只有两个键的键盘/minSteps.go
package main

import (
	"fmt"
	"math"
)

func main() {
	fmt.Println(getSqrt(8))
	fmt.Println(minSteps(8))
}

func minSteps(n int) int {
	if n == 1 {
		return 0
	}
	dp := make([]int, n+1)
	for i := 2; i <= n; i++ {
		dp[i] = i

		for j := 2; j <= getSqrt(n); j++ {
			if i%j == 0 {
				dp[i] = dp[j] + dp[i/j]
				break
			}
		}
	}
	return dp[n]
}

func getSqrt(num int) int {
	return int(math.Ceil(math.Sqrt(float64(num))))
}
